mem/mem.go

```diff
@@ -0,0 +1,325 @@
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package mem
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"sync/atomic"
+)
+
+// DefaultChannelSize is the default Pool channel size.
+//
+// See the WithChannelSize function for more details.
+const DefaultChannelSize uint16 = 4
+
+// DefaultSegListSizes are the default seg list sizes.
+//
+// See the WithSegListSizes function for more details.
+var DefaultSegListSizes = []int{
+	8,
+	16,
+	64,
+	256,
+	4096,
+	65536,
+	1048576,
+	16777216,
+}
+
+var globalPool = NewPool(WithNoLocking(), WithDisabled())
+
+// Global returns the global Pool instance.
+//
+// It is generally preferable to create a Pool instance for your application and
+// pass it through to the places it is needed, however this will prove difficult for
+// existing applications wishing to migrate to the Pool API.
+//
+// The Global Pool is disabled by default. To enable pooling in your application using
+// the Get* generated functions for each message, call Global().Enable() at initialization.
+func Global() *Pool {
+	return globalPool
+}
+
+// Bytes represents a byte slice created from a Pool.
+//
+// Only create these from Pools.
+type Bytes struct {
+	segListPool *segListPool
+	value       []byte
+	valueLen    int
+}
+
+// NewUnmanagedBytes creates a new Bytes that is not backed by any Pool.
+//
+// Users should generally not call this directly.
+func NewUnmanagedBytes(valueLen int) *Bytes {
+	return &Bytes{
+		value:    make([]byte, valueLen),
+		valueLen: valueLen,
+	}
+}
+
+// Value gets the value.
+func (b *Bytes) Value() []byte {
+	return b.value[:b.valueLen]
+}
+
+// Len gets the length.
+//
+// This is equivalent to len(b.Value()).
+func (b *Bytes) Len() int {
+	return b.valueLen
+}
+
+// Truncate truncates the value to the given length.
+func (b *Bytes) Truncate(valueLen int) {
+	b.valueLen = valueLen
+}
+
+// MemsetZero sets the data inside the value to zero.
+//
+// Optimized per https://golang.org/cl/137880043
+func (b *Bytes) MemsetZero() {
```

Comment:
In the above linked CL this is referred to as `MemClr`. Do we want to follow that precedent?